src/redux/sagas/auto-opponent.ts

```diff
@@ -0,0 +1,79 @@
+import {
+  takeEvery,
+  put,
+  take,
+  actionChannel,
+  select,
+} from 'redux-saga/effects';
+import { GameActionType, GameAction, MoveSentAction } from '../actions/game';
+import { fetchOrCreateWallet } from './wallet';
+import { MessageActionType, MessageAction } from '../actions/messages';
+import { fromProposal, GameEngine } from '../../game-engine/GameEngine';
+import Move from '../../game-engine/Move';
+import { WaitForBToDeposit } from '../../game-engine/application-states/PlayerA';
+import { ReadyToChooseBPlay } from '../../game-engine/application-states/PlayerB';
+import { Play } from '../../game-engine/positions';
+import { getUser } from '../store';
+
+export default function* autoOpponentSaga() {
+  yield takeEvery(GameActionType.PLAY_COMPUTER, startAutoOpponent);
+}
+
+function* startAutoOpponent() {
+  const user = yield select(getUser);
+
+  const wallet = yield fetchOrCreateWallet(`AutoPlayer_${user.uid}`);
+  yield put(GameAction.chooseOpponent(wallet.address, 50));
+
+  let gameEngine: GameEngine | null = null;
+
+  const actionFilter = (action): boolean => {
+    return (
+      (action.type === MessageActionType.SEND_MESSAGE &&
+        action.to === wallet.address) ||
+      (action.type === GameActionType.STATE_CHANGED &&
+        action.state instanceof WaitForBToDeposit)
+    );
+  };
+
+  // Get a channel of actions we're interested in
+  const channel = yield actionChannel(actionFilter);
+
+  while (true) {
+    const action = yield take(channel);
+
+    if (gameEngine === null) {
+      // Start up the game engine for our autoplayer B
+      gameEngine = fromProposal({ move: Move.fromHex(action.data), wallet });
+      yield put(MessageAction.messageReceived(gameEngine.state.move.toHex()));
+    } else {
+      switch (action.type) {
+        case MessageActionType.SEND_MESSAGE:
+          yield handleMessage(gameEngine);
+          break;
+
+        case GameActionType.STATE_CHANGED:
```

Comment:
This is a little bit confusing to read, as you have to look back to know that `STATE_CHANGED => state == WaitToDeposit` due to the filter. I'm not sure how to make this clearer though. Maybe add a comment?

Reply:
Good point! We could change the filter to just catch all `STATE_CHANGED` actions and then have an `if` statement in the case. Would make it clearer but we'd be catching a ton more actions. Might just add a comment for now.